source/common/common/thread_synchronizer.cc

```diff
@@ -0,0 +1,92 @@
+#include "common/common/thread_synchronizer.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Thread {
+
+ThreadSynchronizer::~ThreadSynchronizer() {
+  if (data_ != nullptr) {
+    // Make sure we don't have any pending signals which would indicate a bad test.
+    for (auto& entry : data_->entries_) {
```

Comment:
nit: prefer not using `auto` in this scenario.

Moreover you could just do the ASSERT in a one-line destructor in `SynchronizerEntry` and drop the loop altogether; In fact the entire destructor could be dropped.